### PR TITLE
Refactor management of third-parties utilities.

### LINKED ISSUE #128: Move CMake options to configure the tools in a YAML file
# Status

Many variables are defined at CMake level. It makes them a bit complicated to overwrite them (#117), and it is required to run the `configure` phase to use the associated tools (ClangFormat, ...). Furthermore, the CLI utilities are unusable right away since most of their inputs are defined at CMake level.

# Target

* Do not have to run `cmake` to use most of the tools. There are still inherent limits though, for instance:
  * ClangTidy requires:
    * A `compile_commands.json` file (conveniently generated by CMake)
    * All source files to be generated. In NMODL, it means that some make targets must be executed before running the static analysis.
* Centralize the configuration such that it can be exploited by both CMake and the CLI utilities
* Make this configuration easier to maintain and expand

# Proposal

Remove some CMake variables and put them in a new configuration file. hpc-coding-conventions project would provide a default one, but a parent project could provide its own.

## The config file

in YAML format. Filename: `.cppproject.yml`
Default file used by hpc-coding-conventions in `cpp/.cppproject.yml

## List of CMake cache variables to remove

* `${CODING_CONV_PREFIX}_GIT_COMMIT_HOOKS`
* `${CODING_CONV_PREFIX}_GIT_PUSH_HOOKS`
* `${CODING_CONV_PREFIX}_ClangFormat_REQUIRED_VERSION`
* `${CODING_CONV_PREFIX}_ClangFormat_OPTIONS`
* `${CODING_CONV_PREFIX}_ClangFormat_FILES_RE`
* `${CODING_CONV_PREFIX}_ClangFormat_EXCLUDES_RE`
* `${CODING_CONV_PREFIX}_ClangFormat_DEPENDENCIES`
* `${CODING_CONV_PREFIX}_CMakeFormat_REQUIRED_VERSION`
* `${CODING_CONV_PREFIX}_CMakeFormat_OPTIONS`
* `${CODING_CONV_PREFIX}_CMakeFormat_FILES_RE`
* `${CODING_CONV_PREFIX}_CMakeFormat_EXCLUDES_RE`
* `${CODING_CONV_PREFIX}_ClangTidy_REQUIRED_VERSION`
* `${CODING_CONV_PREFIX}_ClangTidy_OPTIONS`
* `${CODING_CONV_PREFIX}_ClangTidy_FILES_RE`
* `${CODING_CONV_PREFIX}_ClangTidy_EXCLUDES_RE`
* `${CODING_CONV_PREFIX}_ClangTidy_DEPENDENCIES`
* `${CODING_CONV_PREFIX}_TEST_FORMATTING`

## Content of the default configuration file

```yaml
tools:

  # a tool is implicitly enabled when specified here

  # not all sections are mandatory, not mentioning a tool means it is not enabled.

  ClangFormat:

    # Any Python version specifiers as described in PEP 440
    # https://peps.python.org/pep-0440/#compatible-release
    # For instance:
    #
    # Exact match required: == 13.0.0
    # At least 13.0.1, but less than 13.1: >= 13.0.1, == 13.0.*
    # Same as above: ~= 13.0.1
    version: ==13.0.0

    # passed to clang-format command line. Can be:
    # - None
    # - a string
    # - a list of string
    option:

    include:
      match:
      - .*\.[it]?cc?$
      - .*\.hh?$
      - .*\.[chit]((pp)|(xx))$

    # targets that needs to be built before formatting the code
    cmake_deps:
    - build_cpp_headers

  CMakeFormat:
    version: ">=0.6, <0.7"
    include:
      match:
      - .*\.cmake$
      - .*CMakeLists.txt$


  ClangTidy:
    # static analysis with ClangTidy is explicitely disabled
    enable: False
    version: ">=7"
    option: -extra-arg=-Wno-unknown-warning-option
    include:
      match: .*\.c((c)|(pp)|(xx))$

  PreCommit:
    enable: False
    hooks:
      commit:
      - check-clang-format
      - check-cmake-format
      push:
      - clang-tidy
      - courtesy-msg

  # Global settings to every tool, can be locally overriden within the
  # section config of a tool

  global:
    exclude:
      match:
      - third[-_]party/.*
      - third[-_]parties/.*
```

### Finding utilities like ClangTidy or CMakeFormat

Using Python version specifiers in the YAML file to describe the versions required is a move toward the idea to create a virtualenv to automagically install the required tools. Right now, the utilities like ClangTidy (an executable) or CMakeFormat (a Python package) are searched by CMake. It is probable that introducing Python version specifiers means the end of this and rely on Python to find (and maybe later install) the required tools.

### PR DESCRIPTION
**Pull request by tristan0x**
This patch introduces:
* 2 new executables:
  * `bin/format`
  * `bin/clang-tidy`
* a YAML configuration file to describe the different tools supported by this project. So far, ClangFormat, CMakeFormat, Black, ClangTidy, and Flake8 are supported.
* `cpp/lib.py` which is the v2 of `cpp/cmake/cpplib.py`.

They are meant to replace (most of) the CMake variables that were used to configure the formatting of C++/CMake files and static analysis of C++ files.

This change also removes a lot of Python and CMake code, not needed anymore.

Mostly fix #128
see touchdetector merge request 31

### CODE REVIEW

**Review by matz-e — COMMENTED** [review]
Posting my mid-review comments for now.

This is exciting, thank you! Ping me when you're done, I'd like to give it another once over if needed.

**Review by matz-e — APPROVED** [review]

File: bin/.bbp-task [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+
+sys.path.append(os.path.dirname(os.path.dirname(os.path.realpath(__file__))))
```
[the remark this answers is not in the document]
Comment (tristan0x) [reply bin/.bbp-task]: Done in [638954c](https://github.com/BlueBrain/hpc-coding-conventions/pull/132/commits/638954cb6fb529055aa68114e323cfa0347b953b)
Thank you for the suggestion

File: bbp-project.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    include:
+      match:
+      - .*\.[it]?cc?$
+      - .*\.hh?$
+      - .*\.[chit]((pp)|(xx))$
+
+  CMakeFormat:
+    enable: False
+    version: ">=0.6, <0.7"
```
[the remark this answers is not in the document]
Comment (tristan0x) [reply bbp-project.yaml]: I wouldn't bet my life on this but it hasn't been the case for the last 2 years. I prefer to give them the benefits of the doubt than freezing the version but I really don't have a strong opinion. I am ok to freeze it if you prefer to be conservative on this.

**Comment by tristan0x** [comment]
> Posting my mid-review comments for now.
> 
> This is exciting, thank you! Ping me when you're done, I'd like to give it another once over if needed.

I am done!
You can merge it and start playing with it in touchdetector, but my next working day is July 4th. I will be around until June 8th to provide support but after that don't count on me ✈️

**Review by matz-e — APPROVED** [review]

File: cpp/lib.py [hunk 3]
```diff
@@ -241,23 +246,23 @@ def bin_dir(self) -> str:
         Return:
             Path to the /bin directory of the virtual environment
         """
-        return os.path.join(self.path, "bin")
+        return self.path.joinpath("bin")
```
[the remark this answers is not in the document]
Comment (1uc) [reply cpp/lib.py:249]: There's many more of these. I don't mind much which version is chosen `.joinpath` look clear to me. I'd suggest consistency though. (One reason to prefer `.joinpath` is that it looks like `os.path.join` which behaves really well for `join("", "foo")`. Unlike what bash does for `"${EMPTY_STRING}/foo"`. The slash notation reminds me of bash.)

File: cpp/lib.py [hunk 4]
```diff
@@ -0,0 +1,1352 @@
...
+        pip_pkg:
+            - `True` if the tool is an installable Python package
+            - the package name if different than the tool
+            - `False` if the tool is not a Python package
+    """
+
+
+class BBPProject:
+    CONFIG_FILE = "bbp-project.yaml"
+    USER_CONFIG_FILE = ".bbp-project.yaml"
```
Comment (1uc) [inline cpp/lib.py:823]:
```suggestion
    # This is the config file for the project being formatted. The path is
    # relative to root of the project using the HPC CC.
    USER_CONFIG_FILE = ".bbp-project.yaml"
```

**Comment by 1uc** [comment]
I ran the newest version for TD. It fails with:
```
$ ./deps/hpc-coding-conventions/bin/format -n 
Traceback (most recent call last):
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/./deps/hpc-coding-conventions/bin/format", line 18, in <module>
    sys.exit(0 if main() == 0 else 1)
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/./deps/hpc-coding-conventions/bin/format", line 14, in main
    return BBPProject.run_task(task)
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1041, in run_task
    project = BBPProject.from_config_files()
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1335, in from_config_files
    cls.virtualenv().ensure_requirement("PyYAML>=5")
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 948, in virtualenv
    return BBPVEnv(source_dir().joinpath(".bbp-project-venv"))
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 53, in source_dir
    return git_dir.parent
AttributeError: 'str' object has no attribute 'parent'
```

File: cpp/lib.py [hunk 5]
```diff
@@ -0,0 +1,1352 @@
...
+        dry_run = kwargs.get("dry_run", False)
+
+        task_config = self.config["provides"][task]
+        if dry_run:
+            cmd = cmd + self.cmd_opts(task, **kwargs) + list(files)
+            log_command(cmd, logger=self.job_logger, level=logging.INFO)
+            status = subprocess.call(cmd, stderr=subprocess.DEVNULL)
+            if status != 0:
+                lang_str = ", ".join(task_config["languages"])
+                logging.error(
```
[the remark this answers is not in the document]
Comment (matz-e) [reply cpp/lib.py:610]: Good point, overlooked that there are colors in there! That trips all kinds of things up.

File: cpp/lib.py [hunk 6]
```diff
@@ -0,0 +1,1352 @@
...
+    TOOLS_DESCRIPTION = dict(
+        ClangFormat=dict(
+            cls=ExecutableTool,
+            name="clang-format",
+            names_glob=["clang-format-*"],
+            version_opt=["--version"],
+            version_re="([0-9]+\\.[0-9]+\\.[0-9]+)",
+            capabilities=ToolCapabilities(
+                cli_accept_dir=False,
+                cli_max_num_files=30,
```
[the remark this answers is not in the document]
Comment (tristan0x) [reply cpp/lib.py:850]: Several reasons:
* There is a kernel limit for a command line size. There are good changes one single clang-format command with all neuron sources specified in argument may exceed this limit
* It allows running these commands in // in a close future, with the `multiprocessing` standard module for instance.


The chosen value is completely arbitrary

File: cpp/lib.py [hunk 7]
```diff
@@ -0,0 +1,1352 @@
...
+
+        Return:
+            instance of argparse.ArgumentParser
+        """
+        task_config = cls.TASKS_DESCRIPTION[task]
+        parser = argparse.ArgumentParser(description=task_config.description)
+        ns = argparse.Namespace()
+        if task_config.modify_files:
+            parser.add_argument(
+                "-n",
```
[the remark this answers is not in the document]
Comment (tristan0x) [reply cpp/lib.py:966]: I have seen `-n` in a few tools, but I can only recall `make`.

File: cpp/lib.py [hunk 8]
```diff
@@ -0,0 +1,1352 @@
...
+
+        task_config = self.config["provides"][task]
+        if dry_run:
+            cmd = cmd + self.cmd_opts(task, **kwargs) + list(files)
+            log_command(cmd, logger=self.job_logger, level=logging.INFO)
+            status = subprocess.call(cmd, stderr=subprocess.DEVNULL)
+            if status != 0:
+                lang_str = ", ".join(task_config["languages"])
+                logging.error(
+                    "\033[1;31merror:\033[0;0m"
```
[the remark this answers is not in the document]
Comment (tristan0x) [reply cpp/lib.py]: Thank you for your remarks. I removed the color and the level repetition here
[0dee53e](https://github.com/BlueBrain/hpc-coding-conventions/pull/132/commits/0dee53e8181c71f68f0ac52efcc27819f134a714)

File: cpp/lib.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tristan0x) [reply cpp/lib.py:249]: Using `pathlib` is already a good improvement. `/` is too fast for me @matz-e  
Besides, operator overloading is evil :)

**Comment by 1uc** [comment]
I've tried to run it in TD. I'm still hitting an issue:
```
$ cd deps/hpc-coding-conventions/ && git rev-parse HEAD && cd -
6348cb78035f97bb6257591f787ac7d964a77ce2

$ module list
Currently Loaded Modulefiles:
  1) unstable               3) git/2.31.1             5) python/3.9.7
  2) spack                  4) cmake/3.21.4           6) py-virtualenv/16.7.6

$ python --version
Python 3.9.7

$ which python
~/touchdetector/.bbp-project-venv/bin/python

$ ./deps/hpc-coding-conventions/bin/format -n 
Traceback (most recent call last):
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/./deps/hpc-coding-conventions/bin/format", line 18, in <module>
    sys.exit(0 if main() == 0 else 1)
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/./deps/hpc-coding-conventions/bin/format", line 14, in main
    return BBPProject.run_task(task)
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1051, in run_task
    num_errors = project.run_task_on_codebase(task, **vars(options))
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1093, in run_task_on_codebase
    [tool.configure() for tool in tools]
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1093, in <listcomp>
    [tool.configure() for tool in tools]
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 650, in configure
    self._path, self._version = self.find_tool_in_path()
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 675, in find_tool_in_path
    all_paths = [(p, self.find_version(p)) for p in paths]
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 675, in <listcomp>
    all_paths = [(p, self.find_version(p)) for p in paths]
  File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 705, in find_version
    return pkg_resources.get_distribution(pkg_name).version
  File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 466, in get_distribution
    dist = get_provider(dist)
  File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 342, in get_provider
    return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
  File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 886, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 772, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'clang-format' distribution was not found and is required by the application
```
Let me provide context. AFAICT, it starts here:
```
            try:
                print(f" -- requirement = {self.requirement}" )
                self._path, self._version = self.find_tool_in_path()   # LINE 650
            except FileNotFoundError as e:
                print(f"requirement = {self.requirement}" )
                if self.requirement:
                    BBPProject.virtualenv().ensure_requirement(self.requirement)
                    self._path, self._version = self.find_tool_in_path(
                        [BBPProject.virtualenv().bin_dir]
                    )
                else:
                    raise e

            # NOTE: any error that isn't a FileNotFoundError will not
            # be caught here.
            logging.info( ... )
```
The trace show that a `pkg_resources.DistributionNotFound` is thrown. This happens when calling  `self.find_version(p)` which calls `pkg_resources.get_distribution(pkg_name).version` (line 705).

Since this exception is caught it doesn't trigger the whole `pip install` workflow.

**Comment by 1uc** [comment]
Let me disable auto-merge.

**Comment by tristan0x** [comment]
> I've tried to run it in TD. I'm still hitting an issue:
> 
> ```
> $ cd deps/hpc-coding-conventions/ && git rev-parse HEAD && cd -
> 6348cb78035f97bb6257591f787ac7d964a77ce2
> 
> $ module list
> Currently Loaded Modulefiles:
>   1) unstable               3) git/2.31.1             5) python/3.9.7
>   2) spack                  4) cmake/3.21.4           6) py-virtualenv/16.7.6
> 
> $ python --version
> Python 3.9.7
> 
> $ which python
> ~/touchdetector/.bbp-project-venv/bin/python
> 
> $ ./deps/hpc-coding-conventions/bin/format -n 
> Traceback (most recent call last):
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/./deps/hpc-coding-conventions/bin/format", line 18, in <module>
>     sys.exit(0 if main() == 0 else 1)
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/./deps/hpc-coding-conventions/bin/format", line 14, in main
>     return BBPProject.run_task(task)
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1051, in run_task
>     num_errors = project.run_task_on_codebase(task, **vars(options))
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1093, in run_task_on_codebase
>     [tool.configure() for tool in tools]
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 1093, in <listcomp>
>     [tool.configure() for tool in tools]
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 650, in configure
>     self._path, self._version = self.find_tool_in_path()
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 675, in find_tool_in_path
>     all_paths = [(p, self.find_version(p)) for p in paths]
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 675, in <listcomp>
>     all_paths = [(p, self.find_version(p)) for p in paths]
>   File "/gpfs/bbp.cscs.ch/home/groshein/touchdetector/deps/hpc-coding-conventions/cpp/lib.py", line 705, in find_version
>     return pkg_resources.get_distribution(pkg_name).version
>   File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 466, in get_distribution
>     dist = get_provider(dist)
>   File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 342, in get_provider
>     return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]
>   File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 886, in require
>     needed = self.resolve(parse_requirements(requirements))
>   File "/gpfs/bbp.cscs.ch/ssd/apps/bsd/2022-01-10/stage_applications/install_gcc-11.2.0-skylake/py-setuptools-57.4.0-v4z3s5/lib/python3.9/site-packages/pkg_resources/__init__.py", line 772, in resolve
>     raise DistributionNotFound(req, requirers)
> pkg_resources.DistributionNotFound: The 'clang-format' distribution was not found and is required by the application
> ```
> 
> Let me provide context. AFAICT, it starts here:
> 
> ```
>             try:
>                 print(f" -- requirement = {self.requirement}" )
>                 self._path, self._version = self.find_tool_in_path()   # LINE 650
>             except FileNotFoundError as e:
>                 print(f"requirement = {self.requirement}" )
>                 if self.requirement:
>                     BBPProject.virtualenv().ensure_requirement(self.requirement)
>                     self._path, self._version = self.find_tool_in_path(
>                         [BBPProject.virtualenv().bin_dir]
>                     )
>                 else:
>                     raise e
> 
>             # NOTE: any error that isn't a FileNotFoundError will not
>             # be caught here.
>             logging.info( ... )
> ```
> 
> The trace show that a `pkg_resources.DistributionNotFound` is thrown. This happens when calling `self.find_version(p)` which calls `pkg_resources.get_distribution(pkg_name).version` (line 705).
> 
> Since this exception is caught it doesn't trigger the whole `pip install` workflow.

Thank you for the feedback @1uc 
It has been fixed in 1c710992ecfed3ae1a8bce0f3cde53ef823713b4

**Comment by tristan0x** [comment]
ready for review @1uc and @matz-e

**Review by 1uc — DISMISSED** [review]
Very excited to start using this. Thank you.

I've checked that it:
- installs the correct version of `clang-format` if needed.
- runs as expected on TD when using the `venv` in `.bbp-project-venv`.
- runs as expected when using an env which has the correct version of the tools installed.

**Comment by tristan0x** [comment]
Thanks for the review and testing @1uc 
I have just pushed your suggestion and tested it on the TouchDetector repository.

**Review by 1uc — APPROVED** [review]

**Pull request merged**